src/files.ml

```diff
@@ -12,52 +217,47 @@ let obj_extension : string = ".lpo"
-
-(** [pp oc mp] prints [mp] to channel [oc]. *)
-let pp_path : module_path pp = fun oc mp ->
-  Format.pp_print_string oc (String.concat "." mp)
-
-(** [module_path path] computes the [module_path] corresponding to a  relative
-    file [path], which should not use [".."]. The returned list is formed with
-    the subdirectories along the [path], and it is terminated by the file name
-    (without extension). Although it is removed, the extension should be given
-    on the file name, and it should correspond to [src_extension] (or possibly
-    to the legacy extension [legacy_src_extension]).  When [path] is  invalid,
-    [Invalid_argument "invalid module path"] is raised. *)
-let module_path : string -> module_path = fun fname ->
-  let ext_ok = Filename.check_suffix fname src_extension in
-  let ext_ok = ext_ok || Filename.check_suffix fname legacy_src_extension in
-  if not ext_ok then
-    fatal_no_pos "Invalid file extension for [%s] (expected [%s] or [%s])."
-      fname src_extension legacy_src_extension;
-  if not (Filename.is_relative fname) then
-    fatal_no_pos "Invalid path for [%s] (expected relative path)." fname;
-  let base = Filename.chop_extension (Filename.basename fname) in
-  let dir = Filename.dirname fname in
-  if dir = Filename.parent_dir_name then
-    fatal_no_pos "Invalid path for [%s] ([%s] not allowed)."
-      fname Filename.parent_dir_name;
-  let rec build_path acc dir =
-    let dirbase = Filename.basename dir in
-    let dirdir  = Filename.dirname  dir in
-    if dirdir = Filename.parent_dir_name then
-      fatal_no_pos "Invalid path for [%s] ([%s] not allowed)."
-        fname Filename.parent_dir_name;
-    if dirbase = Filename.current_dir_name then acc
-    else build_path (dirbase::acc) dirdir
+(** [file_to_module path] computes the module path that corresponds to [path].
+    The file described by [path] is expected to have a valid extension (either
+    [src_extension] or the legacy extension [legacy_src_extension]). If [path]
+    is invalid, the [Fatal] exception is raised. *)
+let file_to_module : string -> Path.t = fun fname ->
+  (* Sanity check: source file extension. *)
+  let ext = Filename.extension fname in
+  if not (List.mem ext [ src_extension ; legacy_src_extension ]) then
+    fatal_no_pos "Invalid extension for [%s] (expected [%s] or [%s])." fname
+      src_extension legacy_src_extension;
+  (* Normalizing the file path. *)
+  let fname =
+    try Filename.realpath fname with Invalid_argument(_) ->
+      fatal_no_pos "No such file or directory [%s]." fname
+  in
+  let base = Filename.chop_extension fname in
+  (* Finding the best mapping under the root. *)
+  let mapping = ref None in
+  let fn mp path =
+    if String.is_prefix path base then
+      match !mapping with
+      | None       -> mapping := Some(mp, path)
+      | Some(_, p) -> if String.length p < String.length path then
+                        mapping := Some(mp, p)
+  in
+  ModMap.iter fn (current_mappings ());
```

Comment:
Can't you define a more efficient ModMap.find function that would not go through all mappings?

Reply:
No, you absolutely need to go through all mappings to find the "best" one.

Reply:
Ok but, then, why not moving this code in the  ModMap module?

Reply:
The `ModMap` module is more of a library on which the function relies. Anyway, I might refactor this code at some point. For now, it is still a work in progress, that reviewing at that level of detail is not useful yet.